go/vt/vtgate/evalengine/eval_result.go

```diff
@@ -253,24 +294,66 @@ func (er *EvalResult) setTuple(t []EvalResult) {
 func (er *EvalResult) makeBinary() {
 	er.resolve()
 	if er.bytes_ == nil {
-		panic("called makeBinary on non-textual")
+		er.bytes_ = er.toRawBytes()
 	}
 	er.type_ = int16(sqltypes.VarBinary)
 	er.collation_ = collationBinary
 	er.clearFlags(flagBit | flagHex)
 }
 
-func (er *EvalResult) clearFlags(flags uint16) {
-	er.flags_ &= ^flags
+func (er *EvalResult) clearFlags(f flag) {
+	er.flags_ &= ^f
 }
 
 func (er *EvalResult) makeTextual(collation collations.ID) {
 	er.resolve()
 	if er.bytes_ == nil {
-		panic("called makeTextual on non-textual")
+		er.bytes_ = er.toRawBytes()
 	}
+	er.collation_.Collation = collation
 	er.type_ = int16(sqltypes.VarChar)
+}
+
+func (er *EvalResult) makeTextualAndConvert(collation collations.ID) bool {
+	er.resolve()
+	if er.bytes_ == nil {
+		er.bytes_ = er.toRawBytes()
+	}
+	if er.collation_.Collation == collations.Unknown {
+		er.collation_.Collation = collations.CollationBinaryID
+	}
+
+	var err error
+	environment := collations.Local()
+	fromCollation := environment.LookupByID(er.collation_.Collation)
+	toCollation := environment.LookupByID(collation)
+	er.bytes_, err = collations.Convert(nil, toCollation, er.bytes_, fromCollation)
+	if err != nil {
+		er.setNull()
+		return false
+	}
+
 	er.collation_.Collation = collation
+	er.type_ = int16(sqltypes.VarChar)
+	return true
+}
+
+func (er *EvalResult) truncate(size int) {
+	switch er.typeof() {
+	case sqltypes.VarBinary:
+		if size > len(er.bytes_) {
+			pad := make([]byte, size)
+			_ = append(pad[:0], er.bytes_...)
+			er.bytes_ = pad
```

Comment:
what does `_ = append(pad[:0], er.bytes_...)` do?

Reply:
It means I brokened this! It was supposed to be a `copy` not an `append`. Gonna fix.